src/tokenpool-gnu-make.cc

```diff
@@ -0,0 +1,140 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "tokenpool-gnu-make.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+#include "line_printer.h"
+
+using namespace std;
+
+// TokenPool implementation for GNU make jobserver - common bits
+// every instance owns an implicit token -> available_ == 1
+GNUmakeTokenPool::GNUmakeTokenPool() : available_(1), used_(0) {
+}
+
+GNUmakeTokenPool::~GNUmakeTokenPool() {
+}
+
+bool GNUmakeTokenPool::SetupClient(bool ignore,
+                                   bool verbose,
+                                   double& max_load_average) {
+  const char* value = GetEnv("MAKEFLAGS");
+  if (!value)
+    return false;
+
+  // GNU make <= 4.1
+  const char* jobserver = strstr(value, "--jobserver-fds=");
+  if (!jobserver)
+    // GNU make => 4.2
+    jobserver = strstr(value, "--jobserver-auth=");
+  if (jobserver) {
+    LinePrinter printer;
+
+    if (ignore) {
+      printer.PrintOnNewLine("ninja: warning: -jN forced on command line; ignoring GNU make jobserver.\n");
+    } else {
+      if (ParseAuth(jobserver)) {
+        const char* l_arg = strstr(value, " -l");
+        int load_limit = -1;
+
+        if (verbose) {
+          printer.PrintOnNewLine("ninja: using GNU make jobserver.\n");
+        }
+
+        // translate GNU make -lN to ninja -lN
+        if (l_arg &&
+            (sscanf(l_arg + 3, "%d ", &load_limit) == 1) &&
+            (load_limit > 0)) {
+          max_load_average = load_limit;
+        }
+
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+bool GNUmakeTokenPool::SetupMaster(bool verbose,
+                                   int parallelism,
+                                   double max_load_average,
+                                   const char* style) {
+  // no need to set up token pool for serial builds
+  if (parallelism == 1)
+    return false;
+
+  string auth;
+  if (CreatePool(parallelism, style, &auth)) {
+    string value = "--jobserver-auth=" + auth;
+    if (max_load_average > 0.0f) {
+      char buffer[32];
+      snprintf(buffer, sizeof(buffer), "%g", max_load_average);
+      value += " -l";
+      value += buffer;
+    }
+
+    if (SetEnv("MAKEFLAGS", value.c_str())) {
+      if (verbose) {
+        LinePrinter printer;
+        printer.PrintOnNewLine("ninja: simulating GNU make jobserver.\n");
+      }
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool GNUmakeTokenPool::Acquire() {
+  if (available_ > 0)
+    return true;
+
+  if (AcquireToken()) {
+    // token acquired
+    available_++;
+    return true;
+  }
+
+  // no token available
+  return false;
+}
+
+void GNUmakeTokenPool::Reserve() {
+  available_--;
+  used_++;
+}
+
+void GNUmakeTokenPool::Return() {
+  if (ReturnToken())
+    available_--;
+}
+
+void GNUmakeTokenPool::Release() {
+  available_++;
+  used_--;
+  if (available_ > 1)
```

Comment:
You shouldn't do that synchronously on every `Release()`. Return tokens in bulk when you start ***blocking*** on sub-process completion, but before that, keep holding on!